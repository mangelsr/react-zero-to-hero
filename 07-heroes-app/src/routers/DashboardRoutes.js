import React from 'react';
import { Redirect, Route, Switch } from 'react-router-dom';

import { DcScreen } from '../components/dc/DcScreen';
import { MarvelScreen } from '../components/marvel/MarvelScreen';
import { SearchScreen } from '../components/search/SearchScreen';
import { HeroScreen } from '../components/ui/heroes/HeroScreen';
import { Navbar } from '../components/ui/Navbar';


export const DashboardRoutes = () => {
    return (
        <>
            <Navbar />
            <div className="container mt-3">
                <Switch>
                    <Route exact path="/dc" component={DcScreen} />
                    <Route exact path="/marvel" component={MarvelScreen} />
                    <Route exact path="/hero/:heroId" component={HeroScreen} />
                    <Route exact path="/search" component={SearchScreen} />
                    <Redirect to="/dc" />
                </Switch>
            </div>
        </>
    )
}
