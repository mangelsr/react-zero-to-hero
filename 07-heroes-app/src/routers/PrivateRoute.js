import React from 'react'
import { Redirect, Route } from 'react-router-dom';
import PropTypes from 'prop-types';


export const PrivateRoute = ({ 
    isAuth,
    component: Component,
    ...rest
}) => {

    localStorage.setItem('lastpath', rest.location.pathname);

    return (
        <Route {...rest}
            component={props => isAuth
                ? <Component {...props} /> 
                : <Redirect to="/login" />}
            >            
        </Route>
    );
}

PrivateRoute.propTypes = {
    isAuth: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
};