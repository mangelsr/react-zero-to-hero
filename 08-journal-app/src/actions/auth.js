import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from "sweetalert2";

import { types } from "../types/types";
import { uiEndLoading, uiStartLoading } from "./ui";
import { clearNotes } from "./notes";

export const startLoginEmailPassword = (email, password) => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().signInWithEmailAndPassword(email, password);
            dispatch(login(user.uid, user.displayName));
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
}

export const startGoogleLogin = () => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().signInWithPopup(googleAuthProvider);
            dispatch(login(user.uid, user.displayName))
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
};


export const startRegister = (email, password, name) => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().createUserWithEmailAndPassword(email, password);
            await user.updateProfile({displayName: name});
            dispatch(login(user.uid, user.displayName));
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const startLogout = () => {
    return async (dispatch) => {
        try {
            await firebase.auth().signOut();
            dispatch(logout());
            dispatch(clearNotes());
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        }
    };
};

export const logout = () => ({
    type: types.logout
});