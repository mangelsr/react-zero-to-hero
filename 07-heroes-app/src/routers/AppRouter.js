import React, { useContext } from 'react';
import {
    BrowserRouter as Router,
    Switch
} from "react-router-dom";

import { AuthContext } from '../auth/AuthContext';
import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';


export const AppRouter = () => {

    const { user:{logged} } = useContext(AuthContext);

    return (
        <div>
            <Router>
            <div>
                <Switch>
                    <PublicRoute isAuth={logged} exact path="/login" component={LoginScreen} />
                    <PrivateRoute isAuth={logged} path="/" component={DashboardRoutes} />
                </Switch>
            </div>
            </Router>
        </div>
    )
}
