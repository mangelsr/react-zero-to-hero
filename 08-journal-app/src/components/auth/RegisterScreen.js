import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from "validator";

import { startRegister } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';


export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { loading, msgError } = useSelector(state => state.ui);

    const [ formValues, handleInputChange ] = useForm({
        name: 'Angel',
        email: 'angel@gmail.com',
        password: '123456',
        password2: '123456',
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = e => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startRegister(email, password, name));
        }
    };

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            return false;
        } else if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            return false;
        } else if (password !== password2 || password.length < 5) {
            dispatch(setError('Password should match and be at least 6 characers'));
            return false;
        }
        dispatch(removeError());
        return true;
    };

    return (
        <div className="animate__animated animate__fadeIn">
            <h3 className="auth__title">Register</h3>
            <form onSubmit ={handleRegister}>

                { msgError && 
                    <div className="auth__alert-error">
                        { msgError }
                    </div>
                }
                
                <input onChange={handleInputChange} value={name} className="auth__input" type="text" placeholder="Name" name="name" autoComplete="off"/>
                <input onChange={handleInputChange} value={email} className="auth__input" type="text" placeholder="Email" name="email" autoComplete="off"/>
                <input onChange={handleInputChange} value={password} className="auth__input" type="password" placeholder="Password" name="password"/>
                <input onChange={handleInputChange} value={password2} className="auth__input" type="password" placeholder="Password confirmation" name="password2"/>
                
                <button disabled={loading} className="btn btn-primary btn-block mb-5" type="submit">Register</button>

                <Link className="link" to="/auth/login">
                    Already registered?
                </Link>
            </form>
        </div>
    );
}
