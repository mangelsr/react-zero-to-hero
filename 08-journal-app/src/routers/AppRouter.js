import React, { useEffect, useState } from 'react';
import { BrowserRouter as Router, Switch, Redirect } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { AuthRouter } from './AuthRouter';
import { JournalScreen } from '../components/journal/JournalScreen';
import { firebase } from "../firebase/firebaseConfig";
import { login } from '../actions/auth';
import { PublicRoute } from './PublicRoute';
import { PrivateRoute } from './PrivateRoute';
import { startLoadingNotes } from '../actions/notes';


export const AppRouter = () => {

    const dispatch = useDispatch();
    const [checking, setChecking] = useState(true);
    const [isLoggedIn, setIsLoggedIn] = useState(false);

    useEffect(() => {
        firebase.auth().onAuthStateChanged(async (user) => {
            if (user?.uid) {
                dispatch(login(user.uid, user.displayName));
                setIsLoggedIn(true);
                dispatch(startLoadingNotes(user.uid)) ;
            } else {
                setIsLoggedIn(false);
            }
            setChecking(false);
        });
    }, [dispatch, setChecking, setIsLoggedIn]);

    if (checking) {
        return <h1>Wait...</h1>;
    } else {
        return (
            <Router>
                <Switch>
                    <PublicRoute isAuth={isLoggedIn} path="/auth" component={AuthRouter} />
                    <PrivateRoute exact isAuth={isLoggedIn} path="/" component={JournalScreen} />
                    <Redirect to="/" />
                </Switch>
            </Router>
        );
    }
}
