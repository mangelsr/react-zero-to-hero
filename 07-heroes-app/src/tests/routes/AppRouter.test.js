import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';

import { AppRouter } from '../../routers/AppRouter';
import { AuthContext } from '../../auth/AuthContext';


describe('Tests over <AppRouter /> component', () => {

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: false
        }
    };
    
    test('should login if not authenticated', () => {
        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <AppRouter />
            </AuthContext.Provider>
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should render DCComponent if is authenticated', () => {
        const contextValue = {
            dispatch: jest.fn(),
            user: {
                logged: true,
                name: 'Miguel'
            }
        }

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <AppRouter />
            </AuthContext.Provider>
        );
        expect(wrapper.find('.navbar').exists()).toBe(true);
    });

});