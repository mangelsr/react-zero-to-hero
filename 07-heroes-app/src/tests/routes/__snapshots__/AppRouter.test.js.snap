// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests over <AppRouter /> component should login if not authenticated 1`] = `
<div>
  <div>
    <div
      className="container mt-5"
    >
      <h1>
        Login Screen
      </h1>
      <hr />
      <button
        className="btn btn-outline-success"
        onClick={[Function]}
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
